File: Year-2025 Problem List/March 2025/2594. Minimum Time to Repair Cars/2594. Minimum Time to Repair Cars.cpp
# Problem          : 2594. Minimum Time to Repair Cars
# Time Complexity  : O(n⋅log(m⋅max_rank))
# C++ Solution

class Solution {
public:
    long long repairCars(vector<int>& ranks, int cars) {
        auto count_repaired = [&](long long time) -> long long {
            long long cnt = 0;
            for (auto rank : ranks) {
                cnt += static_cast<int>(std::sqrt(time / rank));
            }
            return cnt;
        };

        long long left = 0, right = 1ll * ranks[0] * cars * cars;
        long long res = -1;

        while (left <= right) {
            long long mid = left + (right - left) / 2;
            long long repaired = count_repaired(mid);

            if (repaired >= cars) {
                res = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return res;
    }
};

File: Year-2025 Problem List/March 2025/2594. Minimum Time to Repair Cars/2594. Minimum Time to Repair Cars.py
# Problem          : 2594. Minimum Time to Repair Cars
# Time Complexity  : O(n⋅log(m⋅max_rank))
# Python Solution

class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:

        def count_repaired(time: int) -> int:
            cnt = 0
            for rank in ranks:
                cnt += int(math.sqrt(time / rank))
            return cnt

        left, right = 0, ranks[0] * cars * cars
        res = -1

        while left <= right:
            mid = left + (right - left) // 2
            required = count_repaired(mid)

            if required >= cars:
                res = mid
                right = mid - 1
            else:
                left = mid + 1
        return res